sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs

```diff
@@ -82,7 +84,18 @@ public sealed class EventDataBatch : IDisposable
         ///
         /// <returns><c>true</c> if the event was added; otherwise, <c>false</c>.</returns>
         ///
-        public bool TryAdd(EventData eventData) => InnerBatch.TryAdd(eventData);
+        public bool TryAdd(EventData eventData)
+        {
+            bool instrumented = EventDataInstrumentation.InstrumentEvent(eventData, Activity.Current);
```

Comment:
oh, this would mean that all messages in the batch share the same Id. This does not look right - how would you distinguish those? i.e. causation is broken

Reply:
The batch is sent as a single message over the wire, and received by Event Hubs as a single entity.   Each event in that batch is a member, not a stand-alone message.  I'm missing enough overall context around the distributed tracing goals that I'm not able to reconcile whether that addresses your concern or not, but it feels like it would be relevant.

Reply:
> oh, this would mean that all messages in the batch share the same Id. 

Not exactly, every message would have the ID of the Activity it was added within.  I'm not sure how adding a different ID to every message helps causation. Do you have some use case in mind?

Reply:
I think I understand, because there are no other effective message IDs, diagnostic ID is only thing that can uniquly identify the message down the line.

Reply:
But on the other hand even the doc says ` 'Diagnostic-Id' uniquely identifies operation that enqueued the event `. Which doesn't require ID per message.

Reply:
Would one of you be so kind as to define `message` in this context?   It may help me better understand the goal.

Reply:
When I say message I mean event.

Reply:
Is the underlying goal to trace units that correspond to each service interaction or the components of those?   For example, if this were a REST service where we made a POST request, would we be trying to instrument the POST or each item in the payload of that request?

Reply:
@jsquire  the underlying goal is to trace each event (or samples) end-to-end across multiple services. We will have instrumentation for sending and processing (receiving is a different beast and we may tackle it later).

For HTTP we trace all http requests and each of them carries a unique context.

Event through in case of EventHub we batch event together, their origin (and context they-we been created in) is more important than the fact they-ve been sent together.

Here is the more broad spec on this: https://gist.github.com/lmolkova/e4215c0f44a49ef824983382762e6b92

@pakrym so the spec says
> Every message MUST carry trace context and trace context MUST be unique per message if tracing is enabled

> SDK MUST generate context for message when creating message OR when sending it.

> SDK MUST create a span for each send attempt and add links to each message's context (when send span is not related to message context). SDK MUST set links before starting the span.

This way we create Activity/span per message and create unique id). So when there are two messages that are created in scope of one request, you can distinguish them and trace them separately (causation is preserved by them having unqiue context). If they share the same context you are no longer able to tell which one went where and how it was processed.

Reply:
Creating a 0 length span/Activity for the message just to generate an ID feels completely wrong. The message is not an operation it's an object.

Reply:
you also need to log it - not just generate a unique id. And while it might feel strange, this is the only feasible way to go using existing things we have.

We kinda want timestamp, id (and eventually the rest of the context) and we want to log it somewhere, we may even want tags (message-id or other properties). The two things we don't need for it are duration and status.

So I'd argue it's a reasonable thing to do.

Everyone else does it (google in their pub-sub), zipkin recommends for kafka it and community-based kafka instrumentations do it too.